agent/stats/engine.go

```diff
@@ -461,6 +464,15 @@ func (engine *DockerStatsEngine) GetInstanceMetrics(includeServiceConnectStats b
 			ContainerMetrics:      containerMetrics,
 		}
 
+		if includeServiceConnectStats {
+			serviceConnectStats, ok := engine.taskToServiceConnectStats[taskArn]
+			if !ok {
+				seelog.Debugf("task '%s' is not registered to collect service connect metrics", taskArn)
```

Comment:
this looks odd. I wonder if this is an error?

Reply:
hmmm since service connect might not be enabled for all the tasks, thought logging an error wouldn't make sense!

Reply:
but don't we set `includeServiceConnectStats=true` only if `task.IsServiceConnectEnabled()`?

Reply:
currently `includeServiceConnectStats` flag is only used to indicate that we need to publish metrics with service connect stats on every 3rd(or 60s) request -  https://github.com/aws/amazon-ecs-agent/blob/feature/service-connect/agent/tcs/client/client.go#L183-L189

Meanwhile one improvement I can think of to reduce this debug logs here is -  if there are no tasks with service connect enabled from ACS tasks that Agent receives we can set `includeServiceConnectStats` to `false`.

Reply:
why not just pass `includeServiceConnectStats=true` when the counter is `3` AND `task.IsServiceConnectEnabled()`?

Reply:
because `publishMetrics()` is for all the tasks the agent is running, not specific to just one task.

Reply:
I see. Although `taskToServiceConnectStats` only has SC-enabled tasks, right? So think this log is redundant since we already know there won't be/should not be SC-enabled tasks in this map.

https://github.com/aws/amazon-ecs-agent/pull/3195/files#diff-50b5b60cbd2aab1f7353cb44f862f4abd0befe3bec0f9ecff0e1844efe18fc36R368

Reply:
hmmm so currently in `publishMertics()` function we are looping over the `taskArns` from `taskContainerMetricsUnsafe` which may/may not be SC-enabled to send the taskMetrics.
Further, if `includeServiceConnectStats` is true, then for given taskArn we check if it SC-enabled by trying to retrieve it from the `taskToServiceConnectStats` map, if found we found it indicates that SC is enabled for the given taskArn so we gather it's service connect stats else we skip collecting the SC stats and just log it.

Basically we are using `taskToServiceConnectStats` to check if task is SC-enabled or not.

Reply:
> Basically we are using taskToServiceConnectStats to check if task is SC-enabled or not.

It doesn't seem like we check if a task is SC-enabled before adding it to the map? Or am I looking at the wrong place?
https://github.com/aws/amazon-ecs-agent/blob/feature/service-connect/agent/stats/engine.go#L370-L372

Reply:
oh yea...I have updated to check in the this PR(line 368) - https://github.com/aws/amazon-ecs-agent/pull/3195/files#diff-50b5b60cbd2aab1f7353cb44f862f4abd0befe3bec0f9ecff0e1844efe18fc36R368

Reply:
Ohhh LOL I conveniently missed that :)